src/ReinforcementLearningDatasets/src/atari/atari_dataset.jl

```diff
@@ -0,0 +1,169 @@
+"""
+function dataset(game::String,
+    index::Int,
+    epochs::Vector{Int};
+    style=SARTS,
+    repo = "atari-replay-datasets",
+    rng = StableRNG(123), 
+    is_shuffle = true, 
+    batch_size=256
+)
+    
+    try 
+        @datadep_str "$repo-$game-$index"
+    catch
+        throw("The provided dataset is not available") 
+    end
+        
+    path = @datadep_str "$repo-$game-$index" 
+
+    @assert length(readdir(path)) == 1
+    folder_name = readdir(path)[1]
+    
+    folder_path = "$path/$folder_name"
+    files = readdir(folder_path)
+    file_prefixes = map(x->join(split(x,"_")[1:2], "_"), files)
+    set_file_prefixes = collect(Set(file_prefixes))
+    fields = map(collect(set_file_prefixes)) do x
+        if split(x, "_")[1] == "\$store\$"
+            x = split(x, "_")[2]
+        else
+            x = x
+        end
+    end
+
+    s_epochs = Set(epochs)
+    
+    dataset = Dict()
+
+    for (prefix, field) in zip(set_file_prefixes, fields)
+        for epoch in s_epochs
+            data = gzopen("$folder_path/$(prefix)_ckpt.$epoch.gz") do file
+                npzreadarray(file)
+            end
+
+            if field == "observation"
+                data = permutedims(data, [3, 2, 1]) # not sure about the orientation of the frame
+            end
+
+            if haskey(dataset, field)
+                if field == "observation"
+                    dataset[field] = cat(dataset[field], data, dims=3)
+                else
+                    dataset[field] = cat(dataset[field], data, dims=1)
+                end
+            else
+                dataset[field] = data
+            end
+        end
+    end
+
+    num_epochs = length(s_epochs)
+
+    atari_verify(dataset, num_epochs) 
+
+    N_samples = size(dataset["observation"])[3]
+    
+    final_dataset = Dict{Symbol, Any}()
+    meta = Dict{String, Any}()
+
+    for (key, d_key) in zip(["observation", "action", "reward", "terminal"], Symbol.(["state", "action", "reward", "terminal"]))
+            final_dataset[d_key] = dataset[key]
+    end
+    
+    for key in keys(dataset)
+        if !(key in ["observation", "action", "reward", "terminal"])
+            meta[key] = dataset[key]
+        end
+    end
+
+    return AtariDataSet(final_dataset, epochs, repo, N_samples, batch_size, style, rng, meta, is_shuffle)
+
+end
+
+function iterate(ds::AtariDataSet, state = 0)
+    rng = ds.rng
+    batch_size = ds.batch_size
+    size = ds.size
+    is_shuffle = ds.is_shuffle
+    style = ds.style
+
+    if is_shuffle
+        inds = rand(rng, 1:size, batch_size)
+        map((x)-> if x <= size x else 1 end, inds)
+    else
+        if (state+1) * batch_size <= size
+            inds = state*batch_size+1:(state+1)*batch_size
+        else
+            return nothing
+        end
+        state += 1
+    end
+
+    batch = (state = copy(ds.dataset[:state][:, :, inds]),
```

Comment:
I think we don't need copy here?

Reply:
Yep. I just wanted to give the user the flexibility to change the value of the batches.

I haven't come across such a use. Will change it.